builtin/providers/circonus/resource_circonus_rule_set.go

```diff
@@ -0,0 +1,859 @@
+							}
+						}
+					case ruleSetMetricTypeText:
+						if v, found := valueAttrs[ruleSetAbsentAttr]; found {
+							s := v.(string)
+							if s != "" {
+								d, _ := time.ParseDuration(s)
+								rule.Criteria = apiRuleSetAbsent
+								rule.Value = float64(d.Seconds())
+								break METRIC_TYPE
+							}
+						}
+
+						if v, found := valueAttrs[ruleSetChangedAttr]; found {
+							b := v.(bool)
+							if b {
+								rule.Criteria = apiRuleSetChanged
+								break METRIC_TYPE
+							}
+						}
+
+						if v, found := valueAttrs[ruleSetContainsAttr]; found {
+							s := v.(string)
+							if s != "" {
+								rule.Criteria = apiRuleSetContains
+								rule.Value = s
+								break METRIC_TYPE
+							}
+						}
+
+						if v, found := valueAttrs[ruleSetEqualsAttr]; found {
+							s := v.(string)
+							if s != "" {
+								rule.Criteria = apiRuleSetMatch
+								rule.Value = s
+								break METRIC_TYPE
+							}
+						}
+
+						if v, found := valueAttrs[ruleSetExcludesAttr]; found {
+							s := v.(string)
+							if s != "" {
+								rule.Criteria = apiRuleSetNotMatch
+								rule.Value = s
+								break METRIC_TYPE
+							}
+						}
+
+						if v, found := valueAttrs[ruleSetMissingAttr]; found {
+							s := v.(string)
+							if s != "" {
+								rule.Criteria = apiRuleSetNotContains
+								rule.Value = s
+								break METRIC_TYPE
+							}
+						}
+					default:
+						return fmt.Errorf("PROVIDER BUG: unsupported rule set metric type: %q", rs.MetricType)
+					}
+
+					if ruleSetOverListRaw, found := valueAttrs[ruleSetOverAttr]; found {
+						overList := ruleSetOverListRaw.(*schema.Set).List()
+
+						for _, overListRaw := range overList {
+							overAttrs := newInterfaceMap(overListRaw)
+
+							if v, found := overAttrs[ruleSetLastAttr]; found {
+								last, err := time.ParseDuration(v.(string))
+								if err != nil {
+									return errwrap.Wrapf(fmt.Sprintf("unable to parse duration %s attribute", ruleSetLastAttr), err)
+								}
+								rule.WindowingDuration = uint(last.Seconds())
+							}
+
+							if v, found := overAttrs[ruleSetUsingAttr]; found {
+								s := v.(string)
+								rule.WindowingFunction = &s
+							}
+						}
+					}
+				}
+			}
+			rs.Rules = append(rs.Rules, rule)
+		}
+	}
+
+	if v, found := d.GetOk(ruleSetTagsAttr); found {
+		rs.Tags = derefStringList(flattenSet(v.(*schema.Set)))
+	}
+
+	if err := rs.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (rs *circonusRuleSet) Create(ctxt *providerContext) error {
+	crs, err := ctxt.client.CreateRuleSet(&rs.RuleSet)
+	if err != nil {
+		return err
+	}
+
+	rs.CID = crs.CID
+
+	return nil
+}
+
+func (rs *circonusRuleSet) Update(ctxt *providerContext) error {
+	_, err := ctxt.client.UpdateRuleSet(&rs.RuleSet)
+	if err != nil {
+		return errwrap.Wrapf(fmt.Sprintf("Unable to update rule set %s: {{err}}", rs.CID), err)
+	}
+
+	return nil
+}
+
+func (rs *circonusRuleSet) Validate() error {
```

Comment:
If possible I want to leave it as a stub because I know there are a few client-side checks that we could provide to validate the inputs.  For instance:

```
  // For an 'on absence' rule this is the number of seconds the metric must not
  // have been collected for, and should not be lower than either the period or
  // timeout of the metric being collected.
```

As we flex this code path I plan on adding checks here where necessary because we can provide a more rich error message that provides context vs a `400` response code with a sometimes cryptic error message.